crates/bevy_asset/src/loader.rs

```diff
@@ -618,6 +618,90 @@ impl<'a> LoadContext<'a> {
+        &mut self,
+        reader: &mut Reader<'_>,
+        type_id: TypeId,
+    ) -> Result<ErasedLoadedAsset, LoadDirectError> {
+        let loader = self
+            .asset_server
+            .get_asset_loader_with_asset_type_id(type_id)
+            .await
+            .map_err(|error| LoadDirectError {
+                dependency: self.asset_path.clone_owned(),
+                error: error.into(),
+            })?;
+
+        let meta = loader.default_meta();
+
+        let load_context = LoadContext::new(
+            self.asset_server,
+            self.asset_path.clone_owned(),
+            self.should_load_dependencies,
+            self.populate_hashes,
+        );
+
+        let loaded_asset = loader
+            .load(reader, meta, load_context)
+            .await
+            .map_err(|error| LoadDirectError {
+                dependency: self.asset_path.clone_owned(),
+                error: AssetLoadError::AssetLoaderError {
+                    path: self.asset_path.clone_owned(),
+                    loader_name: loader.type_name(),
+                    error,
+                },
+            })?;
+
+        Ok(loaded_asset)
+    }
+
+    /// Loads the asset from the given `reader` directly as the asset type `type_id`. This is an async function that will wait until the asset is fully loaded before
+    /// returning. Use this if you need the _value_ of another asset in order to load the current asset, and that value comes from your [`Reader`].
+    /// For example, if you are deriving a new asset from the referenced asset, or you are building a collection of assets.
+    pub async fn load_direct_with_reader<'b, A: Asset>(
+        &mut self,
+        reader: &mut Reader<'_>,
+    ) -> Result<A, LoadDirectError> {
+        let loader = self
+            .asset_server
+            .get_asset_loader_with_asset_type::<A>()
+            .await
+            .map_err(|error| LoadDirectError {
+                dependency: self.asset_path.clone_owned(),
+                error: error.into(),
+            })?;
+
+        let meta = loader.default_meta();
+
+        let load_context = LoadContext::new(
+            self.asset_server,
+            self.asset_path.clone_owned(),
+            self.should_load_dependencies,
+            self.populate_hashes,
+        );
+
+        let loaded_asset = loader
+            .load(reader, meta, load_context)
+            .await
+            .map_err(|error| LoadDirectError {
+                dependency: self.asset_path.clone_owned(),
+                error: AssetLoadError::AssetLoaderError {
+                    path: self.asset_path.clone_owned(),
+                    loader_name: loader.type_name(),
+                    error,
+                },
+            })?
+            .take::<A>()
+            // AssetServer::get_asset_loader_with_asset_type::<A> can only return an `AssetLoader<Asset = A>`
+            .unwrap();
+
+        Ok(loaded_asset)
+    }
```

Comment:
Doesnt this method do the exact same thing as `load_direct_with_reader_and_type_id` exect with the generic type instead? If that is the case, would it not make more sense to call `load_direct_with_reader_and_type_id` here and not duplicate this code?

Inside this method you call `asset_server.get_asset_loader_with_asset_type` which is just a wrapper around `asset_server.self.get_asset_loader_with_asset_type_id(TypeId::of::<A>())`. The whole method could be reduced to (not tested)
```rs
pub async fn load_direct_with_reader<'b, A: Asset>(reader..){
    self.load_direct_with_reader_and_type(reader, TypeId::of(A))
      .map(|loaded_asset| loaded_asset.take::<A>().unwrap())
}
```

Reply:
I've removed `load_direct_with_reader_and_type_id` as a part of upstreaming the current version of Bevy to get this PR into shape for merging (hopefully) soon. As a future PR, I'll look at adding them back in with appropriate application of DRY.